dbus_next/_private/unmarshaller.py

```diff
@@ -1,307 +1,348 @@
+            type_.children[1]
+        )
 
-        key = self.read_argument(type_.children[0])
-        value = self.read_argument(type_.children[1])
-
-        return key, value
-
-    def read_array(self, type_):
-        self.align(4)
-        array_length = self.read_uint32()
+    def read_array(self, type_: SignatureType):
+        self.offset += -self.offset & 3  # align 4 for the array
+        array_length = self.read_argument(UINT32_SIGNATURE)
 
         child_type = type_.children[0]
-        if child_type.token in 'xtd{(':
+        if child_type.token in "xtd{(":
             # the first alignment is not included in the array size
-            self.align(8)
+            self.offset += -self.offset & 7  # align 8
+
+        if child_type.token == "y":
+            self.offset += array_length
+            return self.buf[self.offset - array_length : self.offset]
 
         beginning_offset = self.offset
 
-        result = None
-        if child_type.token == '{':
-            result = {}
+        if child_type.token == "{":
+            result_dict = {}
             while self.offset - beginning_offset < array_length:
                 key, value = self.read_dict_entry(child_type)
-                result[key] = value
-        elif child_type.token == 'y':
-            o = self.read(array_length)
-            # avoid buffer copies when slicing
-            array_mem_slice = memoryview(self.buf)[o:o + array_length]
-            result = array_mem_slice.tobytes()
-        else:
-            result = []
-            while self.offset - beginning_offset < array_length:
-                result.append(self.read_argument(child_type))
-
-        return result
-
-    def read_argument(self, type_):
-        t = type_.token
-
-        if t not in self.readers:
-            raise Exception(f'dont know how to read yet: "{t}"')
-
-        return self.readers[t](type_)
-
-    def _unmarshall(self):
-        self.offset = 0
-        self.read(16, prefetch=True)
-        self.endian = self.read_byte()
-        if self.endian != LITTLE_ENDIAN and self.endian != BIG_ENDIAN:
-            raise InvalidMessageError('Expecting endianness as the first byte')
-        message_type = MessageType(self.read_byte())
-        flags = MessageFlag(self.read_byte())
-
-        protocol_version = self.read_byte()
-
+                result_dict[key] = value
+            return result_dict
+
+        result_list = []
+        while self.offset - beginning_offset < array_length:
+            result_list.append(self.read_argument(child_type))
+        return result_list
+
+    def read_argument(self, type_: SignatureType) -> Any:
+        """Dispatch to an argument reader or cast/unpack a C type."""
+        token = type_.token
+        reader, ctype, size, struct = self.readers[token]
+        if reader:  # complex type
+            return reader(self, type_)
+        self.offset += size + (-self.offset & (size - 1))  # align
+        if self.can_cast:
```

Comment:
This path is still read hot. We could set the function one per unmarshalling based on can_cast ahead of time to reduce the branching